aptos-move/framework/supra-framework/sources/multisig_voting.spec.move

```diff
@@ -0,0 +1,427 @@
+            simple_map::spec_get(post_metadata, is_multi_step_in_execution_key) == std::bcs::serialize(false)
+        } else {
+            !simple_map::spec_contains_key(post_metadata, is_multi_step_in_execution_key)
+        };
+    }
+
+    spec get_resolution_time_secs {
+        include AbortsIfNotContainProposalID<ProposalType>;
+        let voting_forum = global<VotingForum<ProposalType>>(voting_forum_address);
+        let proposal = table::spec_get(voting_forum.proposals, proposal_id);
+        ensures result == proposal.resolution_time_secs;
+    }
+
+    spec is_multi_step_proposal_in_execution {
+        include AbortsIfNotContainProposalID<ProposalType>;
+        let voting_forum = global<VotingForum<ProposalType>>(voting_forum_address);
+        let proposal = table::spec_get(voting_forum.proposals,proposal_id);
+        aborts_if !std::string::spec_internal_check_utf8(IS_MULTI_STEP_PROPOSAL_IN_EXECUTION_KEY);
+
+        let execution_key = std::string::spec_utf8(IS_MULTI_STEP_PROPOSAL_IN_EXECUTION_KEY);
+        aborts_if !simple_map::spec_contains_key(proposal.metadata,execution_key);
+
+        let is_multi_step_in_execution_key = simple_map::spec_get(proposal.metadata,execution_key);
+        aborts_if !from_bcs::deserializable<bool>(is_multi_step_in_execution_key);
+
+        ensures result == from_bcs::deserialize<bool>(is_multi_step_in_execution_key);
+    }
+
+    spec is_voting_period_over {
+        // use supra_framework::chain_status;
+        // requires chain_status::is_operating();
+        // aborts_if false;
+        ensures result == (timestamp::spec_now_seconds() > proposal.expiration_secs);
+    }
+
+    spec fun spec_get_proposal_expiration_secs<ProposalType: store>(
+        voting_forum_address: address,
+        proposal_id: u64,
+    ): u64 {
+        let voting_forum = global<VotingForum<ProposalType>>(voting_forum_address);
+        let proposal = table::spec_get(voting_forum.proposals, proposal_id);
+        proposal.expiration_secs
+    }
+
+    spec fun spec_get_proposal_state<ProposalType>(
+        voting_forum_address: address,
+        proposal_id: u64,
+        voting_forum: VotingForum<ProposalType>
+    ): u64 {
+        let proposal = table::spec_get(voting_forum.proposals, proposal_id);
+        let voting_closed = spec_is_voting_closed<ProposalType>(voting_forum_address, proposal_id);
+        let proposal_vote_cond = (proposal.yes_votes > proposal.no_votes && proposal.yes_votes + proposal.no_votes >= proposal.min_vote_threshold);
+        if (voting_closed && proposal_vote_cond) {
+            PROPOSAL_STATE_SUCCEEDED
+        } else if (voting_closed && !proposal_vote_cond) {
+            PROPOSAL_STATE_FAILED
+        } else {
+            PROPOSAL_STATE_PENDING
+        }
+    }
+
+    spec fun spec_is_voting_closed<ProposalType: store>(voting_forum_address: address, proposal_id: u64): bool {
+        let voting_forum = global<VotingForum<ProposalType>>(voting_forum_address);
```

Comment:
`aborts_if` is missing?

Reply:
This is a spec function. aborts_if should be specified in the spec for actual function.